### PR TITLE
Feature/submission count in exercise

### PR DESCRIPTION
**Pull request by VitaNuova**

### CODE REVIEW

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ResultResource.java [hunk 1]
```diff
@@ -195,7 +196,18 @@ public ResultResource(ResultRepository resultRepository, Optional<LtiService> lt
         } else {
             results = resultRepository.findEarliestSuccessfulResultsForExercise(exerciseId);
         }
-        return results;
+
+        List<ResultWithSubmissionCount> resultsWithSubmissionCounts = new ArrayList<>();
+        List<Object[]> submissionCounts = resultRepository.findSubmissionCountsForStudents(exerciseId);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/ResultResource.java]: Can you add some comments here? Why do we e.g. have List<Object[]>? This looks a bit weired.

[the remark this answers is not in the document]
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ResultResource.java]: I believe this method still has to return List<Object[]>. 
We need to return only two columns as opposed to the whole Participation object.  There is a Spring Data JPA [solution](https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#projections) to this, but it offers no support for aggregate functions (count in our case). 
At the same time, Hibernate uses List<Object[]> for queries returning selected multiple columns, including queries with aggregate functions. See subsection 15.52 in the following [Hibernate documentation section](https://docs.jboss.org/hibernate/stable/orm/userguide/html_single/Hibernate_User_Guide.html#hql)

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/util/ResultWithSubmissionCount.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
+package de.tum.in.www1.exerciseapp.web.rest.util;
+
+import de.tum.in.www1.exerciseapp.domain.Result;
+
+/**
+ * Class used to return a collection of exercise results together with the
+ * number of submissions (results) for the participation the result belongs to
+ */
+public class ResultWithSubmissionCount {
```
[the remark this answers is not in the document]
Comment (VitaNuova) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/util/ResultWithSubmissionCount.java]: Added a new attribute to the Result class and deleted the new class, please have a look.

File: src/main/java/de/tum/in/www1/exerciseapp/domain/Result.java [hunk 3]
```diff
@@ -44,6 +45,20 @@
     @ManyToOne
     private Participation participation;
 
+    // number of results in the participation this result belongs to
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/domain/Result.java]: Can you please provide a bit more description about this attribute using the official JavaDoc comment notation? 1-2 sentences what it actually means and that it is used for statistics, etc.

File: src/main/java/de/tum/in/www1/exerciseapp/repository/ResultRepository.java [hunk 4]
```diff
@@ -31,6 +31,9 @@
     @Query("select r from Result r where r.completionDate = (select min(rr.completionDate) from Result rr where rr.participation.exercise.id = r.participation.exercise.id and rr.participation.student.id = r.participation.student.id and rr.successful = true) and r.participation.exercise.course.id = :courseId and r.successful = true order by r.completionDate asc")
     List<Result> findEarliestSuccessfulResultsForCourse(@Param("courseId") Long courseId);
 
+    // Hibernate returns List<Object[]> when selecting multiple columns
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/repository/ResultRepository.java]: please also provide a more detailed comment using JavaDoc notation, explain the sql statement and the values in the object array

**Pull request merged**